File: Day06/golang/program.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

func run() int {
	file, err := os.Open("input.txt")
	if err != nil {
		log.Fatal(err)
		return 1
	}
	defer file.Close()

	// Fetch data from file arranged in cols with maps holding a occourences counter
	columns := make(map[int]map[rune]int)
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		for col, char := range scanner.Text() {
			if _, err := columns[col]; !err {
				columns[col] = make(map[rune]int)
			}
			columns[col][char]++
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
		return 1
	}

	// Analyze rows in order to find max and min values in cols
	maxOccurrenceMessage := make([]rune, len(columns))
	minOccurrenceMessage := make([]rune, len(columns))
	for colIndex, col := range columns {
		var max int
		var min int
		for key, value := range col {
			if value > max {
				maxOccurrenceMessage[colIndex] = key
				max = value
			}
			if min == 0 || value < min {
				minOccurrenceMessage[colIndex] = key
				min = value
			}
		}
	}

	fmt.Println("Error-corrected version using most occurrences: ", string(maxOccurrenceMessage))
	fmt.Println("Error-corrected version using least occurrences: ", string(minOccurrenceMessage))

	return 0
}

func main() {
	fmt.Println("+-------------------------+")
	fmt.Println("| Advent of Code - Day 06 |")
	fmt.Println("+-------------------------+")

	result := run()

	fmt.Println("  - Glædelig jul!")

	os.Exit(result)
}

File: Day06/DotNet/Program.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

namespace Day06.ConsoleApplication
{
    public class Program
    { 
        public static void Main(string[] args)
        {
            Console.WriteLine("+-------------------------+");
            Console.WriteLine("| Advent of Code - Day 06 |");
            Console.WriteLine("+-------------------------+");
            Console.WriteLine();

            // LINQ only solution (yea, I know... One thing took the other and suddenly the solution was all LINQ'fied)
            var result = Enumerable
                .Range(0, 8)
                .Select(col =>
                    new {
                        col = col,
                        chars = File.ReadAllLines("input.txt").Select(x => x[col])
                    }
                )
                .Select(col =>
                    col.chars.GroupBy(u => u)
                             .Select(g => new { letter = g.Key, count = g.Count()})
                             .OrderByDescending(o => o.count)
                )
                .Select(col =>
                    new {
                        max = col.First().letter.ToString(),
                        min = col.Last().letter.ToString()
                    }
                )
                .Aggregate((a, b) =>
                    new {
                        max = a.max + b.max,
                        min = a.min + b.min
                    }
                );

            Console.WriteLine($"Message using most occurences  : {result.max}");
            Console.WriteLine($"Message using least occurences : {result.min}");
            Console.WriteLine();

            // End of LINQ solution and start of readable solution for the faint hearted
            var input = File.ReadAllLines("input.txt");
            var data = input.First().Select(x => new Dictionary<char, int>()).ToList();

            foreach(var line in input)
            {
                for(int x=0; x<line.Length; ++x)
                {
                    var dict = data[x];
                    var cha = line[x];
                    if(dict.ContainsKey(cha))
                    {
                        ++dict[cha];
                    } else {
                        dict.Add(cha, 1);
                    }
                }
            }

            var most = new StringBuilder();
            var least = new StringBuilder();

            foreach(var col in data)
            {
                var ordered = col.OrderByDescending(x => x.Value);
                most.Append(ordered.First().Key);
                least.Append(ordered.Last().Key);
            }

            Console.WriteLine($"Most is {most} and least is {least}");

            Console.WriteLine($"  -Glædelig jul!");
        }
    }
}